matminer/featurizers/function.py

```diff
@@ -0,0 +1,240 @@
+from __future__ import division
+
+"""
+This module includes code to featurize data based on
+common mathematical expressions used as features.
+
+Note that the approach here has opted for a sympy-based
+parsing of string expressions, rather than explicit
+python functions.  The primary reason this has been
+done is to provide for better support for book-keeping
+(e. g. with feature labels), substitution, and elimination
+of symbolic redundancy, which sympy is well-suited for.
+"""
+
+import numpy as np
+from sympy.parsing.sympy_parser import parse_expr
+import sympy as sp
+import itertools
+from six import string_types
+
+from collections import OrderedDict
+
+from matminer.featurizers.base import BaseFeaturizer
+
+
+# Default expressions to include in function featurizer
+default_exps = ["x", "1/x", "sqrt(x)", "1/sqrt(x)", "x**2", "x**-2", "x**3",
+                "x**-3", "log(x)", "1/log(x)", "exp(x)", "exp(-x)"]
+
+
+# TODO: feature labels might be latexified
+# TODO: parallelization of substitution, think this might be
+# TODO:     unnecessary because featurization already parallelized
+# TODO:     over rows
+class FunctionFeaturizer(BaseFeaturizer):
+    """
+    This class featurizes a dataframe according to a set
+    of expressions representing functions to apply to
+    existing features
+    """
+
+    def __init__(self, expressions=None, multi_feature_depth=1,
+                 postprocess=None, combo_function=None,
+                 latexify_labels=False):
+
+        """
+        Args:
+            expressions ([str]): list of sympy-parseable expressions
+                representing a function of a single variable x, e. g.
+                ["1 / x", "x ** 2"], defaults to the list above
+            multi_feature_depth (int): how many features to include if using
+                multiple fields for functionalization, e. g. 2 will
+                include pairwise combined features
+            postprocess (function or type): type to cast functional outputs
+                to, if, for example, you want to include the possibility of
+                complex numbers in your outputs, use postprocess=np.complex,
+                defaults to float
+            combo_function (function): function to combine multi-features,
+                defaults to np.prod (i.e. cumulative product of expressions),
+                note that a combo function must cleanly process sympy
+                expressions
+            latexify_labels (bool): whether to render labels in latex,
+                defaults to False
+
+        """
+        self.expressions = expressions or default_exps
+        self.multi_feature_depth = multi_feature_depth
+        self.combo_function = combo_function or np.prod
+        self.latexify_labels = latexify_labels
+        self.postprocess = postprocess or float
+
+        # Generate lists of sympy expressions keyed by number of features
+        self.exp_dict = OrderedDict(
```

Comment:
It seems problematic to have this in the constructor (e.g., what if someone changes the list of expressions or multifeature depth after creation?).

That said, I'm not sure of where else to put it. It seems costly to put it inside `_exp_iter`. Any thoughts?

Reply:
I made exp_dict a property so it gets evaluated on the fly (also the combo_function wasn't getting passed to the expression generator, so that's fixed as well).  I don't think it'll be too costly unless someone decides they want to use like 50 functions or a really extreme multi-feature depth, and at that point the majority of the cost will be in iterating over the combinations of input features anyway.  It should be simple to add a lru_cached method for generation, if we find that performance is an issue, but I'd opt for simplicity at this point.